File: backend/types/stack.go
package types

import "database/sql"

type Stack struct {
	Base
	Name         string         `db:"name"`
	PathToFile   string         `db:"path_to_file"`
	Webhook      sql.NullString `db:"webhook"`
	UserID       int            `db:"user_id"`
	RepositoryID sql.NullInt32  `db:"repository_id"`
	Branch       sql.NullString `db:"branch"`
	Roles        sql.NullString `db:"roles"`
}

type StackData struct {
	Name       string `json:"name" validate:"required"`
	Webhook    string `json:"webhook"`
	Repo       string `json:"repo" validate:"required_without=Content"`
	RepoToken  string `json:"repo_token"`
	FileInRepo string `json:"file_in_repo"`
	Content    string `json:"content" validate:"required_without=Repo"`
}

File: frontend/src/types/stack.ts
export class StackFileData {
  name: string;
  content: string;
  webhook: string;
  repo: string;
  repo_token: string;
  file_in_repo: string;

  constructor() {
    this.name = "";
    this.content = "";
    this.webhook = "";
    this.repo = "";
    this.repo_token = "";
    this.file_in_repo = "";
  }
}

export class Stack {
  Id: number;
  CreatedAt: Date;
  UpdatedAt: Date;
  Name: string;
  PathToFile: string;
  Webhook: string | null;
  UserId: number;
  RepositoryId: number | null;
  Branch: string | null;

  constructor(
    id: number,
    createdAt: Date,
    updatedAt: Date,
    name: string,
    pathToFile: string,
    webhook: string | null,
    userId: number,
    repositoryId: number | null,
    branch: string | null
  ) {
    this.Id = id;
    this.CreatedAt = createdAt;
    this.UpdatedAt = updatedAt;
    this.Name = name;
    this.PathToFile = pathToFile;
    this.Webhook = webhook;
    this.UserId = userId;
    this.RepositoryId = repositoryId;
    this.Branch = branch;
  }
}
